### DISCUSSION TITLE
[RFC] virtual file system & NextJS pages as NPM libraries

### DISCUSSION THREAD

**Discussion by eddyw** [vercel/next.js#18778] [Ideas]
# Summary

- provide a way to generate an intermediate build that allows publishing NextJS pages and api routes as an NPM module. For example, `target: "lib"` in `next.config.js`

- provide a virtual file system that allows to consume NextJS NPM modules built with `target: "lib"` and load pages & api routes virtually as if they were part of the app.

For example, in your NextJS app:
```javascript
// next.config.js
modules.exports = {
   virtualFileSystem: ['@my-org/dashboard'],
}
```
And this would generate virtual pages & api routes:
- pages
  - dashboard
    - index.js -> `@my-org/dashboard/build/pages/dashboard/index.js`
    - login.js -> `@my-org/dashboard/build/pages/dashboard/login.js`
    - users
       - [id].js -> `@my-org/dashboard/build/pages/dashboard/users/[id].js`

# Background

It's currently not possible to easily create re-usable pages & api routes across multiple projects. An example of use case is a dashboard built in NextJS that has its own pages & api routes that we want to be able to "mount" in other NextJS apps. Currently, the way to do so is to manually create all api & page routes and import/re-export from the NPM module. For instance:
```javascript
// dashboard/index.js
import { Page, getStaticProps } from '@my-org/dashboard/pages/index`

export { Page as default, getStaticProps }
```

While this approach works, it requires to know the exact directory structure, then re-create it, and then re-export from the library. However, if any of the exports method change (e.g: from `getStaticProps` to `getServerSideProps`) or if the directory structure changes, then it requires to manually do those changes over again in all projects that depend on this dashboard.

# Proposal

### Target "lib"

NextJS should allow to set `target: "lib"` in `next.config.js` which will generate unoptimized code without using env vars, other config, or triggering `get*Props` at build time, and generate pages & api routes in `.next`. This project can then be published to the NPM registry and later be consumable by a NextJS virtual file system.

### Virtual file system

NextJS should allow to set a `virtualFileSystem` array in `next.config.js` that can contain a list of NextJS NPM modules than can be "mounted" or more specifically tell NextJS to load `pages` & `api` routes from an NPM module instead of the local file system. The virtual `pages` & `api` routes would be built using the host `next.config.js` config.

This could provide several benefits:

- Re-use of `pages` & `api` routes across multiple projects
- "Mount" other NextJS projects conditionally based on environment. For instance, a dev-only dashboard when `NODE_ENV === "development"`
- Remove the need for using custom server and get all the benefits from serverless
- inherit env vars and config

A more realistic example, a headless CMS that could provide its dashboard as an NPM module and allow the user to implement their own pages but also manage it.
```javascript
// next.config.js
modules.exports = {
   virtualFileSystem: [
      process.env.NODE_ENV === 'development' && '@my-org/database-dashboard', // dashboard/*
      '@my-org/cms-dashboard', // blog/dashboard
      '@my-org/cms-sitemap-generator', // api/sitemap
   ],
   async rewrites () {
      return [
         {source: '/sitemap.xml', destination: '/api/sitemap'},
      ]
   },
}
```

**Comment by harshvitra** [comment]
@eddyw

Can you provide with a guide on how did you convert your Next js project into a package ?
I am just starting on this and hence need guidance

**Comment by codker** [comment]
I would really like to have this 👍🏻

Reusing pages and APIs is a common problem while using Next.js in big projects.

**Comment by thescientist13** [comment]
Definitely a ➕ 1 to a virtual file system and this idea in general.   👍 

For example, we want to provide SSO middleware to all our apps and don't want every team to have to create a route for handling the SSO callback URL, e.g. `/_sso-provider/token`.  We want to be able to use middleware (currently using a custom _server.js_) to act on and intercept these routes.

However unless there is _actually_ a page for it in the filesystem, e.g.
```js
src/
  pages/
     _sso-provider/
       token.jsx
```

Then Next will always return a `404` status code to the browser, which for one our use cases, would overwrite the `301` status code our middleware adds, thus breaking the browser's automatic `Location` header handling.

But just for SSO alone we have a handful of routes we want to "package" up as middleware, and the same goes for things like Health Check endpoints and other internal / operational related responsibilities of our apps.

----

For now, we just stub these pages out like so which at least gets us a `200`, plus whatever our middleware is doing, but this would require manual intervention by every team consuming our middleware.
```jsx
// src/pages/_sso-provider/[handler].jsx
export default function SsoHandler() {
  return <> </>;
}
```
